usr/src/cmd/bhyve/block_if.c

```diff
@@ -364,9 +364,35 @@ blockif_proc(struct blockif_ctxt *bc, struct blockif_elem *be, uint8_t *buf)
 			else
 				br->br_resid = 0;
 		}
-#endif
 		else
 			 err = EOPNOTSUPP;
+#else
+		else if (bc->bc_ischr) {
```

Comment:
Should this path be taken if it is a block device too?  Or is `F_FREESP` the right path for block devices?

Reply:
There appears to be an implicit assumption that bhyve will either use a character device or a regular file (though sadly, it doesn't appear this is actually enforced anywhere). E.g., `struct blockif_ctx` only contains a flag for 'is a character device' (and not character/block/regular file), and we only do `DKIOCGETWCE` / `DKIOCSETWCE` for character devices.

I'm not sure how bhyve would behave in general if given a block device instead of a character device.  I'd imagine for this the free requests would just fail -- but then the F_{GET,SET}FL calls will fail as well.  Guest _should_ already be prepared for a such an occurrence anyway, and they would already be fairly well outside of 'supported configuration' territory.

I'd be happy to create a followup ticket though that'd enforce (or at least warn) if someone has given bhyve a block device instead of a character device.

Reply:
A low-priority follow-up is fine.